### PR TITLE
Check types in GitHub Actions

### PR DESCRIPTION
**Pull request by masonmcelvain**
Adds a GitHub action that runs the typescript compiler in the frontend, the backend, and every package with a tsconfig.json.

qa_req 0

CC @dhmacs @hcutrone

Connects #695

### CODE REVIEW

**Comment by masonmcelvain** [comment]
Those look like real type failures in our project. I'm not very familiar with the code in `packages/ui/footer/guide`, so I'm inclined to merge this and fix the type failures in a follow up pull. Merging this as-is will illuminate existing type failures in our project, it won't introduce new ones.

**Comment by hcutrone** [comment]
@masonmcelvain The code in `packages/ui/footer/guide` is not being used anywhere (in either repo). They are components I updated to be used in the new guide footer, but ended up not being used. I only kept them because they restyled the footer to match the new design (see description of [this pull](https://github.com/iFixit/react-commerce/pull/678)), so I thought they might be helpful when we updated the collections footer. 

All this to say, it might not be worth it to fix those type bugs.

**Comment by masonmcelvain** [comment]
Thanks for the explanation @hcutrone. I agree that it's probably not worth fixing the type errors right now.

In order for this new workflow to pass on `main` (and all other branches) though, perhaps we can temporarily delete the code that is causing `tsc` errors and then reintroduce it when we update the collections footer later on. Since it's in git, we'll still have a copy of the code, but it won't cause distracting CI failures on other features. What do you think?

**Comment by hcutrone** [comment]
Go for it! I only kept it since it would be easier than searching git

**Comment by masonmcelvain** [comment]
I opted to fix the type errors in a separate pull so that we don't have to block this one on fixing the type errors: https://github.com/iFixit/react-commerce/pull/738

File: .github/workflows/tsc.yml
Comment (dhmacs) [inline .github/workflows/tsc.yml:18]: I think we can just use the npm script, which is also the one that we run on git hook.

```suggestion
           pnpm type-check
```

You can extend it to check the backend folder using the same approach used for the dev script (`npm-run-all`)

**Review by dhmacs — CHANGES_REQUESTED** [review]
Thanks for working on this 👍 I've commented on how to simplify this before we merge

File: .github/workflows/tsc.yml
Comment (masonmcelvain) [reply .github/workflows/tsc.yml]: Good idea, thanks for the suggestion.

I noticed that `pnpm type-check:frontend` overlooked the type errors in `@ifixit/ui` that are being fixed here https://github.com/iFixit/react-commerce/pull/738. Other times, it catches the type errors in `packages/`. Perhaps it's only checking code that is being imported from `packages/` to `frontend/`?

This motivated me to explicitly run `tsc` on all packages to always catch type errors (see https://github.com/iFixit/react-commerce/pull/733/commits/12374962062d76059e34ad191cedf30f420a5f8c). However, this can result in running `tsc` on the same code more than once, which slows things down. What are your thoughts on this?

Comment (dhmacs) [reply .github/workflows/tsc.yml]: @masonmcelvain Yeah I was actually surprised that type checking the frontend wasn't type checking also the dependent packages. I think your approach of type-checking every package is the way to go, and I've found a "more native" pnpm way to do it using this command:

```sh
pnpm run -r type-check
```

The `-r` flag tells pnpm to run the `type-check` script of every app or package defined in `pnpm-workspace.yaml`.
So it's sufficient to declare the `type-check` script on every package, we don't need to define the glue scripts ourselves.
Notably `backend` is not specified as part of the monorepo (due to issues installing Strapi dependencies as part of the monorepo), and that's fine we can just write an additional script for that

**Review by dhmacs — APPROVED** [review]
lgtm CR 👏

**Pull request merged**
